libtenzir/src/tql2/parser.cpp

```diff
@@ -488,15 +510,80 @@ class parser {
     };
   }
 
+  auto token_to_str(token_kind kind) -> std::optional<constant> {
+    TENZIR_ASSERT(kind == tk::string || kind == tk::raw_string);
+    const auto token = accept(kind);
+    if (not token) {
+      return std::nullopt;
+    }
+    auto result = std::string{};
+    if (kind == tk::raw_string) {
+      const auto begin = token.text.find('"');
+      const auto end = token.text.rfind('"');
+      TENZIR_ASSERT(begin != std::string_view::npos);
+      TENZIR_ASSERT(end != std::string_view::npos);
+      result = std::string{token.text.substr(begin + 1, end - begin - 1)};
+    } else if (kind == tk::string) {
```

Comment:
Quick tip: You can improve this function by weakening its preconditions and strengthening its return-type. 

Put differently: You function wants to assume `kind == tk::string || kind == tk::raw_string`. This places an obligation on the caller, which most callers will need to explicitly check to work around (as you did in `parse_record_item`). You could make it `auto accept_string() -> std::optional<...>` instead, so that there are no preconditions to check.  Secondly, your function says "I might have a constant for you", while it could be "I might have a string for you". Because you unnecessarily weakened it, you have to `std::get` the result. If the return type would be the stronger `std::optional<located<std::string>>`, then you would not need to. You can do the conversion to constant in `accept_constant()` instead, which also allows you to remove `parse_string` and `parse_string_raw` (which again have a unnecessary strong precondition that requires checking the token kind before).